// angular-flash-message/dist/flash-message.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
Object.defineProperty(exports, "__esModule", { value: true });
var core_1 = require("@angular/core");
var Subject_1 = require("rxjs/Subject");
var FlashMessage = (function () {
    function FlashMessage() {
        this._successCall = new Subject_1.Subject();
        this.methodCalled = this._successCall.asObservable();
    }
    FlashMessage.prototype.success = function (message, params) {
        if (params === void 0) { params = {}; }
        return this._successCall.next({ message: message, params: params, method: 'success' });
    };
    FlashMessage.prototype.info = function (message, params) {
        if (params === void 0) { params = {}; }
        return this._successCall.next({ message: message, params: params, method: 'info' });
    };
    FlashMessage.prototype.warning = function (message, params) {
        if (params === void 0) { params = {}; }
        return this._successCall.next({ message: message, params: params, method: 'warning' });
    };
    FlashMessage.prototype.danger = function (message, params) {
        if (params === void 0) { params = {}; }
        return this._successCall.next({ message: message, params: params, method: 'danger' });
    };
    FlashMessage = __decorate([
        core_1.Injectable()
    ], FlashMessage);
    return FlashMessage;
}());
exports.FlashMessage = FlashMessage;
