// angular-flash-message/dist/flash-message.component.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
Object.defineProperty(exports, "__esModule", { value: true });
var core_1 = require("@angular/core");
var router_1 = require("@angular/router");
var flash_message_item_1 = require("./flash-message-item");
var flash_message_1 = require("./flash-message");
var FlashMessageComponent = (function () {
    /**
     *
     * @param _flashMessage
     * @param _router
     */
    function FlashMessageComponent(_flashMessage, _router) {
        var _this = this;
        this._flashMessage = _flashMessage;
        this._router = _router;
        this.message = null;
        // window timeout for active message
        this._messageTimeout = null;
        this._publicApiSubscriber = this._flashMessage.methodCalled.subscribe(function (data) {
            var message = data.message, params = data.params, method = data.method;
            _this[method](message, params);
        });
    }
    FlashMessageComponent.prototype.ngOnDestroy = function () {
        this._publicApiSubscriber.unsubscribe();
    };
    /**
     *
     * @param message
     * @param params
     */
    FlashMessageComponent.prototype.success = function (message, params) {
        var _this = this;
        window.clearTimeout(this._messageTimeout);
        var itemParams = this.getItemParams('success', params);
        this.message = new flash_message_item_1.FlashMessageItem(message, itemParams.generalClass, itemParams.close, itemParams.closeBtnClass, itemParams.navigate);
        if (itemParams.navigate && this._router.url !== itemParams.navigate) {
            this._router.navigate([itemParams.navigate]);
        }
        if (this.message.close && !params.delay) {
            return;
        }
        this._messageTimeout = window.setTimeout(function () {
            _this.deleteExistMessage();
        }, itemParams.delay);
    };
    /**
     *
     * @param message
     * @param params
     */
    FlashMessageComponent.prototype.info = function (message, params) {
        var _this = this;
        window.clearTimeout(this._messageTimeout);
        var itemParams = this.getItemParams('info', params);
        this.message = new flash_message_item_1.FlashMessageItem(message, itemParams.generalClass, itemParams.close, itemParams.closeBtnClass, itemParams.navigate);
        if (itemParams.navigate && this._router.url !== itemParams.navigate) {
            this._router.navigate([itemParams.navigate]);
        }
        if (this.message.close && !params.delay) {
            return;
        }
        this._messageTimeout = window.setTimeout(function () {
            _this.deleteExistMessage();
        }, itemParams.delay);
    };
    /**
     *
     * @param message
     * @param params
     */
    FlashMessageComponent.prototype.warning = function (message, params) {
        var _this = this;
        window.clearTimeout(this._messageTimeout);
        var itemParams = this.getItemParams('warning', params);
        this.message = new flash_message_item_1.FlashMessageItem(message, itemParams.generalClass, itemParams.close, itemParams.closeBtnClass, itemParams.navigate);
        if (itemParams.navigate && this._router.url !== itemParams.navigate) {
            this._router.navigate([itemParams.navigate]);
        }
        if (this.message.close && !params.delay) {
            return;
        }
        this._messageTimeout = window.setTimeout(function () {
            _this.deleteExistMessage();
        }, itemParams.delay);
    };
    /**
     *
     * @param message
     * @param params
     */
    FlashMessageComponent.prototype.danger = function (message, params) {
        var _this = this;
        window.clearTimeout(this._messageTimeout);
        var itemParams = this.getItemParams('danger', params);
        this.message = new flash_message_item_1.FlashMessageItem(message, itemParams.generalClass, itemParams.close, itemParams.closeBtnClass, itemParams.navigate);
        if (itemParams.navigate && this._router.url !== itemParams.navigate) {
            this._router.navigate([itemParams.navigate]);
        }
        if (this.message.close && !params.delay) {
            return;
        }
        this._messageTimeout = window.setTimeout(function () {
            _this.deleteExistMessage();
        }, itemParams.delay);
    };
    FlashMessageComponent.prototype.onCloseMessage = function () {
        window.clearTimeout(this._messageTimeout);
        this.deleteExistMessage();
    };
    /**
     * Private component methods
     */
    /**
     *  Delete Existing message
     */
    FlashMessageComponent.prototype.deleteExistMessage = function () {
        this.message = null;
    };
    /**
     *
     * @param messageType
     * @param userParams
     * @returns {{delay: number, generalClass: string}}
     */
    FlashMessageComponent.prototype.getItemParams = function (messageType, userParams) {
        var itemParams = {
            delay: this._delay || 1500,
            generalClass: '',
            close: this._close || false,
            closeBtnClass: this._closeBtnClass || 'flash-message__close-btn',
            navigate: false
        };
        switch (messageType) {
            case 'success':
                itemParams.generalClass = this._successClass || 'alert alert-success';
                break;
            case 'info':
                itemParams.generalClass = this._infoClass || 'alert alert-info';
                break;
            case 'warning':
                itemParams.generalClass = this._warningClass || 'alert alert-warning';
                break;
            case 'danger':
                itemParams.generalClass = this._dangerClass || 'alert alert-danger';
                break;
        }
        for (var optionName in userParams) {
            itemParams[optionName] = userParams[optionName];
        }
        return itemParams;
    };
    __decorate([
        core_1.Input(),
        __metadata("design:type", Number)
    ], FlashMessageComponent.prototype, "_delay", void 0);
    __decorate([
        core_1.Input(),
        __metadata("design:type", String)
    ], FlashMessageComponent.prototype, "_successClass", void 0);
    __decorate([
        core_1.Input(),
        __metadata("design:type", String)
    ], FlashMessageComponent.prototype, "_infoClass", void 0);
    __decorate([
        core_1.Input(),
        __metadata("design:type", String)
    ], FlashMessageComponent.prototype, "_warningClass", void 0);
    __decorate([
        core_1.Input(),
        __metadata("design:type", String)
    ], FlashMessageComponent.prototype, "_dangerClass", void 0);
    __decorate([
        core_1.Input(),
        __metadata("design:type", String)
    ], FlashMessageComponent.prototype, "_close", void 0);
    __decorate([
        core_1.Input(),
        __metadata("design:type", String)
    ], FlashMessageComponent.prototype, "_closeBtnClass", void 0);
    FlashMessageComponent = __decorate([
        core_1.Component({
            selector: 'flash-message',
            template: "\n\t\t<div class=\"{{message.generalClass}}\" *ngIf=\"message\">\n\t\t\t<span>\n\t\t\t\t{{message.message}}\n\t\t\t</span>\n\n\t\t\t<button *ngIf=\"message.close\" class=\"{{message.closeBtnClass}}\" (click)=\"onCloseMessage()\">\n\t\t\t\t<span>X</span>\n\t\t\t</button>\n\t\t</div>\n\t",
            styles: ["\n\t\t.flash-message__close-btn {\n\t\t\tfloat: right;\n\t\t\tborder: none;\n\t\t\tbackground: transparent;\n\t\t\tmargin: 0;\n\t\t\tpadding: 0;\n\t\t}\n\t"]
        }),
        __metadata("design:paramtypes", [flash_message_1.FlashMessage, router_1.Router])
    ], FlashMessageComponent);
    return FlashMessageComponent;
}());
exports.FlashMessageComponent = FlashMessageComponent;
